### PR TITLE
Add write_ipc to ExecutionContext

### LINKED ISSUE #1777: [EPIC] Improve DataFusions ability to work with files
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

I would like to add functionality for writing files from DataFusion.  To start, I've thought of the below.

- [ ] (1) Add write functionality to `ObjectStore`
- [x] (2)Add `write_json` to `ExecutionContext`
- [ ] (3) Add `write_arrow` to `ExecutionContext`
- [x] (4) Add `COPY` / `COPY TO` command for SQL (like postgres https://www.postgresql.org/docs/current/sql-copy.html)
- [ ] (5) Add ability to write partitioned datasets
- [ ] (6) Add support for writing metadata
- [x] (7) Add `write_csv` method to `DataFrame`
- [x] (8) Add `write_parquet` method to `DataFrame`
- [ ] (9) Add `write_arrow` method to `DataFrame`
- [x] (10) Add `write_json` method to `DataFrame`
- [x] (11) Add `read_json` method to `DataFrame`
- [x] (12) Add `read_json` method to `ExecutionContext`
- [x] (13) Add `register_json` method to `ExecutionContext`
- [x] (14) Support `IF NOT EXISTS` for `CREATE EXTERNAL TABLE`
- [ ] (15) Add `read_arrow` to `ExecutionContext`
- [ ] (16) Add `read_arrow` to `DataFrame`

I will use this as a parent / tracker issue for the above points which will each have an issue.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by matthewmturner**
# Which issue does this PR close?

Closes #1777 task 3

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by matthewmturner** [comment]
I'm wondering if `IpcWriteOptions` should implement `clone` and/or `copy` like parquet's `WriterProperties`

File: datafusion/src/execution/context.rs [hunk 1]
```diff
@@ -795,6 +796,57 @@ impl ExecutionContext {
...
+        let path = path.as_ref();
+        // create directory to contain the Parquet files (one per partition)
+        let fs_path = Path::new(path);
+        let runtime = self.runtime_env();
+        match fs::create_dir(fs_path) {
+            Ok(()) => {
+                let mut tasks = vec![];
+                for i in 0..plan.output_partitioning().partition_count() {
+                    let plan = plan.clone();
+                    let filename = format!("part-{}.parquet", i);
```
Comment (alamb) [inline datafusion/src/execution/context.rs:815]: probably needs to be named something other than `.parquet`

File: datafusion/src/execution/context.rs [hunk 2]
```diff
@@ -795,6 +796,57 @@ impl ExecutionContext {
...
+    ) -> Result<()> {
+        let path = path.as_ref();
+        // create directory to contain the Parquet files (one per partition)
+        let fs_path = Path::new(path);
+        let runtime = self.runtime_env();
+        match fs::create_dir(fs_path) {
+            Ok(()) => {
+                let mut tasks = vec![];
+                for i in 0..plan.output_partitioning().partition_count() {
+                    let plan = plan.clone();
```
Comment (alamb) [inline datafusion/src/execution/context.rs:814]: why does the plan need to be cloned?

**Review by alamb — COMMENTED** [review]
We should probably include some tests as well

File: datafusion/src/execution/context.rs [hunk 3]
```diff
@@ -795,6 +794,56 @@ impl ExecutionContext {
...
+        let runtime = self.runtime_env();
+        match fs::create_dir(fs_path) {
+            Ok(()) => {
+                let mut tasks = vec![];
+                for i in 0..plan.output_partitioning().partition_count() {
+                    let filename = format!("part-{}.arrow", i);
+                    let path = fs_path.join(&filename);
+                    let file = fs::File::create(path)?;
+                    let mut writer = match writer_properties {
+                        Some(props) => FileWriter::try_new_with_options(
```
[the remark this answers is not in the document]
Comment (matthewmturner) [reply datafusion/src/execution/context.rs:816]: just to confirm - you would then add the `IpcWriteOptions` as a new parameter to `try_new`, right?

[the remark this answers is not in the document]
Comment (xudong963) [reply datafusion/src/execution/context.rs:816]: Yes, `Option<IpcWriteOptions>`. Then we can `xxx.unwrap_or_else(|| IpcWriteOptions::default())` to process it.

**Comment by alamb** [comment]
> [](https://github.com/matthewmturner)I think IpcWriteOptions should be clone - so im going to submit PR for that in arrow.

👍

**Comment by alamb** [comment]
https://github.com/apache/arrow-datafusion/pull/2048 has the arrow upgrade

**Comment by alamb** [comment]
Marking as a draft to make it clear this still has some work to do. Please mark it ready for review when it is :)

**Comment by alamb** [comment]
This PR is more than 6 month old, so closing it down for now to clean up the PR list. Please reopen if this is a mistake and you plan to work on it more

**Pull request closed without merging**